dashboard/app/tree.go

```diff
@@ -0,0 +1,737 @@
+	if err != nil {
+		return false, err
+	}
+	// It does happen that we sometimes update the tested tree.
+	// It's not frequent at all, but it will make all results very confusing.
+	return build.KernelRepo == mgrBuild.KernelRepo && build.KernelBranch == mgrBuild.KernelBranch, nil
+}
+
+func (test *BugTreeTest) applyPending(c context.Context) error {
+	if test.Pending == "" {
+		return nil
+	}
+	job, err := fetchJob(c, test.Pending)
+	if err != nil {
+		return err
+	}
+	if job.Finished.IsZero() {
+		// Not yet ready.
+		return nil
+	}
+	pendingKey := test.Pending
+	test.Pending = ""
+	if job.Error != 0 {
+		test.Error = pendingKey
+		return nil
+	}
+	test.Last = pendingKey
+	if test.First == "" {
+		test.First = pendingKey
+	}
+	if test.FirstOK == "" && job.CrashTitle == "" {
+		test.FirstOK = pendingKey
+	} else if test.FirstCrash == "" && job.CrashTitle != "" {
+		test.FirstCrash = pendingKey
+	}
+	return nil
+}
+
+func fetchJob(c context.Context, key string) (*Job, error) {
+	jobKey, err := db.DecodeKey(key)
+	if err != nil {
+		return nil, err
+	}
+	job := new(Job)
+	if err := db.Get(c, jobKey, job); err != nil {
+		return nil, fmt.Errorf("failed to get job: %v", err)
+	}
+	return job, nil
+}
+
+type repoNode struct {
+	repo  KernelRepo
+	edges []repoEdge
+}
+
+type repoEdge struct {
+	in    bool
+	merge bool
+	other *repoNode
+}
+
+type repoGraph struct {
+	nodes map[string]*repoNode
+}
+
+func makeRepoGraph(repos []KernelRepo) (*repoGraph, error) {
+	g := &repoGraph{
+		nodes: map[string]*repoNode{},
+	}
+	for _, repo := range repos {
+		if repo.Alias == "" {
+			return nil, fmt.Errorf("one of the repos has an empty alias")
+		}
+		g.nodes[repo.Alias] = &repoNode{repo: repo}
+	}
+	for _, repo := range repos {
+		for _, link := range repo.CommitInflow {
+			if g.nodes[link.Alias] == nil {
+				return nil, fmt.Errorf("no repo with alias %q", link.Alias)
+			}
+			g.nodes[repo.Alias].addEdge(true, link.Merge, g.nodes[link.Alias])
+			g.nodes[link.Alias].addEdge(false, link.Merge, g.nodes[repo.Alias])
+		}
+	}
+	for alias, node := range g.nodes {
+		reachable := node.reachable(true)
+		if _, ok := reachable[node]; ok {
+			return nil, fmt.Errorf("%q lies on a cycle", alias)
+		}
+	}
+	return g, nil
+}
+
+func (g *repoGraph) nodeByRepo(url, branch string) *repoNode {
+	for _, node := range g.nodes {
+		if node.repo.URL == url && node.repo.Branch == branch {
+			return node
+		}
+	}
+	return nil
+}
+
+func (g *repoGraph) nodeByAlias(alias string) *repoNode {
+	for _, node := range g.nodes {
+		if node.repo.Alias == alias {
+			return node
+		}
+	}
+	return nil
+}
+
+// reachable returns a map *repoNode -> bool (whether commits are merged).
+func (n *repoNode) reachable(in bool) map[*repoNode]bool {
+	ret := map[*repoNode]bool{}
+	var dfs func(*repoNode, bool)
+	dfs = func(node *repoNode, merge bool) {
+		for _, edge := range node.edges {
+			if edge.in != in {
+				continue
+			}
+			if _, ok := ret[edge.other]; ok {
+				continue
```

Comment:
Is it that the same tree transitively reachable for both merge=true and merge=false? I think it's possible and then merged=true must take precedence.
It may be done with 2 passes where first traverses just the merged=true part of the graph; or if we recurse if the old value was merged=false, but not we have merged=true.

Reply:
Good catch! I'm not sure it will ever happen in our syzbot configurations, but it's still a problem.

I've fixed it and added a test.